File: funcional/list_compression.hs
import System.CPUTime
import Control.DeepSeq
import Data.List

-- Função para realizar a filtragem dos dados
filtrarDados :: [[String]] -> Int -> String -> [[String]]
filtrarDados dados coluna valor = filter (\linha -> linha !! coluna == valor) dados

-- Função para realizar o mapeamento dos dados
mapearDados :: [[String]] -> Int -> (String -> String) -> [[String]]
mapearDados dados coluna funcao = map (\linha -> [if i == coluna then funcao (linha !! coluna) else valor | (i, valor) <- zip [0..] linha]) dados

-- Função para reduzir os dados
reduzirDados :: [[String]] -> Int -> (Float -> Float -> Float) -> Float
reduzirDados dados coluna funcao = foldl (\acc linha -> funcao acc (read (linha !! coluna)) ) (read (head (dados !! 1)) !! coluna) (tail dados)

-- Função para ordenar os dados
ordenarDados :: [[String]] -> Int -> [[String]]
ordenarDados dados coluna = sortBy (\a b -> compare (a !! coluna) (b !! coluna)) dados

-- Função para carregar os dados de um arquivo CSV
carregarDadosArquivo :: FilePath -> IO [[String]]
carregarDadosArquivo nomeArquivo = do
    conteudo <- readFile nomeArquivo
    let linhas = map (splitOn ',') (lines conteudo)
    return linhas

-- Função auxiliar para dividir uma string em uma lista de strings usando um delimitador
splitOn :: Char -> String -> [String]
splitOn delimiter = foldr f [""] where
    f c l@(x:xs) | c == delimiter = "":l
                 | otherwise = (c:x):xs

-- Exemplo de uso do código
main :: IO ()
main = do
    t1_start <- getCPUTime
    dados <- carregarDadosArquivo "dados.csv" -- UTILIZE QUALQUER CSV que se adeque com o problema.
    t1_stop <- getCPUTime
    let dados_filtrados = filtrarDados dados 1 "30"
    
    t2_start <- getCPUTime
    let dados_mapeados = mapearDados dados_filtrados 2 (\x -> show (read x * 1.1))
    t2_stop <- getCPUTime
    
    t3_start <- getCPUTime
    let media_salarios = reduzirDados dados_mapeados 2 (+) / fromIntegral (length dados_mapeados)
    t3_stop <- getCPUTime
    
    t4_start <- getCPUTime
    let dados_ordenados = ordenarDados dados_mapeados 0
    t4_stop <- getCPUTime

    putStrLn "Dados Filtrados:"
    deepseq dados_filtrados (print dados_filtrados)
    
    putStrLn "Dados Mapeados:"
    deepseq dados_mapeados (print dados_mapeados)
    
    putStrLn "Média dos Salários:"
    deepseq media_salarios (print media_salarios)
    
    putStrLn "Dados Ordenados:"
    deepseq dados_ordenados (print dados_ordenados)
    
    putStrLn $ "Tempo de leitura do arquivo: " ++ show (fromIntegral (t1_stop - t1_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de filtragem dos dados: " ++ show (fromIntegral (t2_stop - t2_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de mapeamento dos dados: " ++ show (fromIntegral (t3_stop - t3_start) / (10^12 :: Double)) ++ " segundos"
    putStrLn $ "Tempo de ordenação dos dados: " ++ show (fromIntegral (t4_stop - t4_start) / (10^12 :: Double)) ++ " segundos"

File: funcional/list_compression.py
from time import process_time
import sys 
sys.setrecursionlimit(10**6) 
# Função para realizar a filtragem recursiva dos dados
def filtrar_dados(dados, coluna, valor):
    return [linha for linha in dados if linha[coluna] == valor]

# Função para realizar o mapeamento dos dados
def mapear_dados(dados, coluna, funcao):
    return [[funcao(linha[coluna]) if i == coluna else linha[i] for i in range(len(linha))] for linha in dados]

def reduzir_dados(dados, coluna, funcao):
    resultado = dados[0][coluna]
    for linha in dados[1:]:
        resultado = funcao(resultado, linha[coluna])
    return resultado

# Função para ordenar os dados
def ordenar_dados(dados, coluna):
    return sorted(dados, key=lambda x: x[coluna])

# Função auxiliar recursiva para ler as linhas do arquivo
def ler_linhas_arquivo(arquivo):
    return [linha.strip().split(",") for linha in arquivo if linha.strip()]

# Função para carregar os dados de um arquivo CSV
def carregar_dados_arquivo(nome_arquivo):
    with open(nome_arquivo, "r") as arquivo:
        dados = ler_linhas_arquivo(arquivo)
    return dados

# Exemplo de uso do código
t1_start = process_time()
dados = carregar_dados_arquivo("dados.csv") # UTILIZE QUALQUER CSV que se adeque com o problema.
t1_stop = process_time()

# Filtrar os dados com base na coluna "idade"
t2_start = process_time()
dados_filtrados = filtrar_dados(dados, coluna=1, valor="30")
t2_stop = process_time()
print("Dados Filtrados:")
print(dados_filtrados)

# Mapear os dados na coluna "salario" com uma função de aumento
t3_start = process_time()
dados_mapeados = mapear_dados(dados_filtrados, coluna=2, funcao=lambda x: float(x) * 1.1)
t3_stop = process_time()
print("Dados Mapeados:")
print(dados_mapeados)

# Reduzir os dados para obter a média dos salários
t4_start = process_time()
media_salarios = reduzir_dados(dados_mapeados, coluna=2, funcao=lambda x, y: x + y) / len(dados_mapeados)
t4_stop = process_time()
print("Média dos Salários:")
print(media_salarios)

# Ordenar os dados pela coluna "nome"
t5_start = process_time()
dados_ordenados = ordenar_dados(dados_mapeados, coluna=0)
t5_stop = process_time()

print("Tempo de leitura do arquivo: ", t1_stop - t1_start)
print("Tempo de filtragem dos dados: ", t2_stop - t2_start)
print("Tempo de mapeamento dos dados: ", t3_stop - t3_start)
print("Tempo de redução dos dados: ", t4_stop - t4_start)
print("Tempo de ordenação dos dados: ", t5_stop - t5_start)
